### PR TITLE
Add authentication against basic auth

### PR DESCRIPTION
**Pull request by nerdmaennchen**
I've added code to authenticate against http basic access authentication.
In my setup I have a (self made) single sign on that can handle basic auth requests (because they are so insanely easy to implement) and use that as an identity and authentication provider.
This PR contains the code that runs on my setup.

### CODE REVIEW

**Comment by violoncelloCH** [comment]
Hi, thank you for your contribution!
DCO is failing because the commit is not signed-off
@nerdmaennchen could you please sign your commit? - see https://github.com/nextcloud/server/blob/master/.github/CONTRIBUTING.md#sign-your-work
can be done with `git commit --amend -s` for the last commit and then `git push -f` to push the changed commit back to github...

**Comment by nerdmaennchen** [comment]
sorry; I didn't know signing was needed.
I've updated the commits and hope its working now.

Cheers,
Lutz

**Comment by violoncelloCH** [comment]
no problem... thank you for adding the sign-off!
we should probably add an info text to the pull request template about that...

File: lib/basicauth.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		  'http'=>array(
+		    'method'=>"GET",
+		    'header' => "authorization: Basic " . base64_encode("$uid:$password")
+		  ))
+		);
+		$headers = get_headers($this->authUrl);
+
+		if($headers === false) {
+			OC::$server->getLogger()->error(
+				'ERROR: Not possible to connect to BasicAuth Url: "'.$this->authUrl.'"',
```
Comment (violoncelloCH) [inline lib/basicauth.php]: this should work without the doubled quotes: `[...] BasicAuth Url: ' . $this->authUrl,`

File: lib/basicauth.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	public function checkPassword($uid, $password) {
+		stream_context_set_default(array(
+		  'http'=>array(
+		    'method'=>"GET",
+		    'header' => "authorization: Basic " . base64_encode("$uid:$password")
+		  ))
+		);
+		$headers = get_headers($this->authUrl);
+
+		if($headers === false) {
```
Comment (violoncelloCH) [inline lib/basicauth.php]: `if (!$headers) {` would be enough here

**Comment by violoncelloCH** [comment]
btw, did you get the unittest you wrote to work? If so, I would appreciate your help at #31, where I'm trying to get the automated tests to work, but I'm stuck there, because I don't have any experience with unittesting yet (I'm just a student and started maintaining this app here two months ago). :smile:

**Comment by nerdmaennchen** [comment]
I'm afraid I didn't get the unittests to work... 
I actually am no php developer at all and most of my code here is copy pasted from multiple sources. 
Also I have this code running on my productive server (which I hope counts as test).

I am, however, very thankful of your assistance and review!

All the best,
Lutz

**Comment by violoncelloCH** [comment]
Ahh, I see... Thank you anyways!

File: README.md [hunk 3]
```diff
@@ -133,6 +133,27 @@ Add the following to your `config.php`:
...
+HTTP server of your choice to authenticate. It should return HTTP 2xx for correct credentials and an appropriate other error code for wrong ones or refused access.
+
+### Configuration
+The only supported parameter is the URL of the web server where the authentication happens.
+
+Add the following to your `config.php`:
+
+    'user_backends' => array(
+        array(
+            'class' => 'OCA_User_BasicAuth',
```
Comment (violoncelloCH) [inline README.md]: `OC_User_BasicAuth`, no `A`...

**Comment by violoncelloCH** [comment]
@MorrisJobke @nickvergessen @rullzer could somebody of you review the unit tests?

**Comment by MorrisJobke** [comment]
> @MorrisJobke @nickvergessen @rullzer could somebody of you review the unit tests?

They look good 👍

File: README.md (hunk 3 above)
Comment (nerdmaennchen) [reply README.md]: ouch.

I blushed quite a lot just now.

**Comment by violoncelloCH** [comment]
@MorrisJobke thank you for reviewing!

**Comment by violoncelloCH** [comment]
@nerdmaennchen looks like you reverted the other corrections while force pushing your last commit... no hurry, just revert it back to what we want :) 
ask if you have any question...

**Comment by nerdmaennchen** [comment]
*blushing intensifies*

I feel like I do owe you a drink.

I hope it is resolved now.

All the best,
Lutz

**Comment by violoncelloCH** [comment]
> I feel like I do owe you a drink.
I hope it is resolved now.

no problem ;) looks like everything is fine now! Thank you for your contribution!

**Comment by nerdmaennchen** [comment]
> > I feel like I do owe you a drink.
> > I hope it is resolved now.
> 
> no problem ;) looks like everything is fine now! Thank you for your contribution!

Well, thank you for developing and maintainig user_external in the first place.
I feel like you deserve a lot of credit as you made it quite easy for dummies like me to add code to this.

**Comment by violoncelloCH** [comment]
> Well, thank you for developing and maintainig user_external in the first place.
I feel like you deserve a lot of credit as you made it quite easy for dummies like me to add code to this.

I did not create it ;)... I'm just maintaining it now as it was unmaintained the months before...

**Comment by jancborchardt** [comment]
@nerdmaennchen no one’s a dummie, welcome to Nextcloud and the @nextcloud/user_external team! :) Since you seem to be in Berlin, be welcome at our Nextcloud Conference (at TU Berlin) in September: https://nextcloud.com/conf-2019/

**Comment by nerdmaennchen** [comment]
Many thanks for the warm welcome!
I'd love to attend to the conference but at the moment (and probably till next spring) I'm on an extended period of traveling (making a life long wish come true). However, every now and then I do have time to spare to do some coding.

I figure the conference happenes every year; I'm looking forward to beeing there next year!

All the best,
Lutz

**Comment by violoncelloCH** [comment]
thank you @jancborchardt for inviting @nerdmaennchen :)

@nerdmaennchen now that you are a member, I wan't to point out to you that you now should be able to push branches directly to Nextcloud's repos (so no need to make a detour via your fork any more) and that you can review and approve PRs ;)

**Pull request merged**